### PR TITLE
Adds support for dynamic dispatch in evaluation

### PR DESCRIPTION
**Pull request by johnedquinn**
## Relevant Issues
- None

## Description
- Adds support for dynamic dispatch in evaluation
- As all of the logic for matching function resides in the planner, I updated the plan representation of dynamic dispatch to hold the expected argument types for each candidate function. This should make our implementation faster, as the checks are minute.
- As the function pathing is still being worked on, I added a temporary deprecated API to help load in functions for execution. This will be removed once pathing is added.

## Other Information
- Updated Unreleased Section in CHANGELOG: **NO**
- Any backward-incompatible changes? **YES**
  - The modelling of dynamic dispatch has changed slightly in the plan representation.
- Any new external dependencies? **NO**
- Do your changes comply with the [Contributing Guidelines](https://github.com/partiql/partiql-lang-kotlin/blob/main/CONTRIBUTING.md)
  and [Code Style Guidelines](https://github.com/partiql/partiql-lang-kotlin/blob/main/CODE_STYLE.md)? **YES**

## License Information

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: partiql-eval/src/main/kotlin/org/partiql/eval/internal/operator/rex/ExprCallDynamic.kt [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    /**
+     * This represents a single candidate for dynamic dispatch.
+     *
+     * This implementation assumes that the [eval] input [Record] contains the original arguments for the desired [fn].
+     * It performs the coercions (if necessary) before computing the result.
+     *
+     * @see ExprCallDynamic
+     */
+    internal class Candidate @OptIn(PartiQLValueExperimental::class, PartiQLFunctionExperimental::class) constructor(
+        val types: Array<PartiQLValueType>,
```
Comment (yliuuuu) [inline partiql-eval/src/main/kotlin/org/partiql/eval/internal/operator/rex/ExprCallDynamic.kt:45]:
```suggestion
        val parameters: Array<PartiQLValueType>,
```

File: partiql-eval/src/main/kotlin/org/partiql/eval/internal/operator/rex/ExprCallDynamic.kt [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        }
+
+        @OptIn(PartiQLValueExperimental::class)
+        internal fun matches(args: Array<PartiQLValue>): Boolean {
+            for (i in args.indices) {
+                if (args[i].type != types[i]) {
+                    return false
+                }
+            }
+            return true
```
Comment (yliuuuu) [inline partiql-eval/src/main/kotlin/org/partiql/eval/internal/operator/rex/ExprCallDynamic.kt:65]: Can you explain this a little more? 

I read the code as for each arg if it is not the original input type, then no match. 

The problem is, after a cast, the actual args will always be the casted type if the cast succeed.

File: partiql-eval/src/main/kotlin/org/partiql/eval/internal/operator/rex/ExprCallDynamic.kt [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+ * arguments, and pass the [PartiQLValue]s directly to the [Candidate.eval].
+ */
+internal class ExprCallDynamic(
+    private val candidates: List<Candidate>,
+    private val args: Array<Operator.Expr>
+) : Operator.Expr {
+
+    @OptIn(PartiQLValueExperimental::class)
+    override fun eval(record: Record): PartiQLValue {
+        val actualArgs = args.map { it.eval(record) }.toTypedArray()
```
Comment (yliuuuu) [inline partiql-eval/src/main/kotlin/org/partiql/eval/internal/operator/rex/ExprCallDynamic.kt:27]: say you entry point is `unionOf(int, decimal) + unionOf(int, decimal)`

You candidates looks like the following: 

```
├──Candidate[tag=Plan-d55fb882]
            │  └──Fn[signature=PLUS__DECIMAL_ARBITRARY_DECIMAL_ARBITRARY__DECIMAL_ARBITRARY, tag=Plan-288ba71]
            ├──Candidate[tag=Plan-e871d747]
            │  ├──Fn[signature=PLUS__DECIMAL_ARBITRARY_DECIMAL_ARBITRARY__DECIMAL_ARBITRARY, tag=Plan-fa4547c]
            │  └──Fn[signature=CAST_DECIMAL_ARBITRARY__INT32__DECIMAL_ARBITRARY, tag=Plan-381675d7]
            ├──Candidate[tag=Plan-52c5cf2b]
            │  ├──Fn[signature=PLUS__DECIMAL_ARBITRARY_DECIMAL_ARBITRARY__DECIMAL_ARBITRARY, tag=Plan-d7c0d7bb]
            │  └──Fn[signature=CAST_DECIMAL_ARBITRARY__INT32__DECIMAL_ARBITRARY, tag=Plan-c86b5dbe]
            └──Candidate[tag=Plan-dfc42937]
               └──Fn[signature=PLUS__INT32_INT32__INT32, tag=Plan-dac73796]
```

Seems to me that one you complete the evaluation of args, then you only need to search the right function to call by grouping the unique function signature.

File: partiql-eval/src/main/kotlin/org/partiql/eval/internal/Compiler.kt [hunk 4]
```diff
@@ -123,6 +129,33 @@ internal class Compiler(
...
+        return ExprCallStatic(function, args)
+    }
+
+    @OptIn(PartiQLFunctionExperimental::class, PartiQLValueExperimental::class)
+    override fun visitRexOpCallDynamic(node: Rex.Op.Call.Dynamic, ctx: Unit): Operator {
+        val args = node.args.map { visitRex(it, ctx) }.toTypedArray()
+        val candidates = node.candidates.map { candidate ->
+            val fn = getFunction(candidate.fn.signature)
+            val coercions = candidate.coercions.map { it?.signature?.let { sig -> getFunction(sig) } }
+            ExprCallDynamic.Candidate(candidate.parameters.toTypedArray(), fn, coercions)
```
Comment (yliuuuu) [inline partiql-eval/src/main/kotlin/org/partiql/eval/internal/Compiler.kt:145]: Mental notes: Parameters maintains the original arg type. So 1 + 1.0 => Plus__DECIMAL__DECIMAL__DECIMAL with Parameter INT and DECIMAL.

File: partiql-eval/src/main/kotlin/org/partiql/eval/internal/operator/rex/ExprCallDynamic.kt [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        val fn: PartiQLFunction.Scalar,
+        val coercions: List<PartiQLFunction.Scalar?>
+    ) {
+
+        @OptIn(PartiQLValueExperimental::class, PartiQLFunctionExperimental::class)
+        fun eval(originalArgs: Array<PartiQLValue>): PartiQLValue {
+            val args = coercions.mapIndexed { index, coercion ->
+                coercion?.invoke(arrayOf(originalArgs[index])) ?: originalArgs[index]
+            }.toTypedArray()
+            return fn.invoke(args)
```
Comment (yliuuuu) [inline partiql-eval/src/main/kotlin/org/partiql/eval/internal/operator/rex/ExprCallDynamic.kt:55]: If a coercion failed, why would we still invoke the function?

File: partiql-eval/src/main/kotlin/org/partiql/eval/internal/operator/rex/ExprCallDynamic.kt (hunk 3 above)
Comment (yliuuuu) [reply partiql-eval/src/main/kotlin/org/partiql/eval/internal/operator/rex/ExprCallDynamic.kt:27]: I see that the coerce function has not been placed in the arg. My bad.

File: partiql-eval/src/main/kotlin/org/partiql/eval/internal/operator/rex/ExprCallDynamic.kt (hunk 5 above)
Comment (johnedquinn) [reply partiql-eval/src/main/kotlin/org/partiql/eval/internal/operator/rex/ExprCallDynamic.kt:55]: If a coercion is absent (AKA NULL), then we use the `originalArgs[index]`. If not, we invoke the coercion.

File: partiql-eval/src/main/kotlin/org/partiql/eval/internal/operator/rex/ExprCallDynamic.kt [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+    @OptIn(PartiQLValueExperimental::class)
+    override fun eval(record: Record): PartiQLValue {
+        val actualArgs = args.map { it.eval(record) }.toTypedArray()
+        candidates.forEach { candidate ->
+            if (candidate.matches(actualArgs)) {
+                candidate.eval(actualArgs)
+            }
+        }
+        return missingValue()
```
[the remark this answers is not in the document]
Comment (johnedquinn) [reply partiql-eval/src/main/kotlin/org/partiql/eval/internal/operator/rex/ExprCallDynamic.kt:33]: In this situation, we need to return missing:
> All functions return MISSING when they input data whose types do not conform to the input argument
types.
>
> -- PartiQL Spec, Section 7.1

So, if a function hasn't been resolved to any of the candidates, we need to return missing.

**Review by yliuuuu — APPROVED** [review]
LGTM

**Pull request merged**